src/Sentry.Profiling/SamplingTransactionProfiler.cs

```diff
@@ -0,0 +1,285 @@
+    private readonly SentryOptions _options;
+
+    public SamplingTransactionProfiler(string tempDirectoryPath, DateTimeOffset now, int timeoutMs, SentryOptions options, CancellationToken cancellationToken)
+    {
+        _options = options;
+        _tempDirectoryPath = tempDirectoryPath;
+        _startTime = now;
+        _cancellationToken = cancellationToken;
+        _session = new(cancellationToken);
+        var clock = Stopwatch.StartNew();
+        _cancellationToken.Register(() =>
+        {
+            if (Stop(now + clock.Elapsed))
+            {
+                _options.LogDebug("Profiling cancelled.");
+            }
+        });
+        Task.Delay(timeoutMs, _cancellationToken).ContinueWith(_ =>
+        {
+            if (Stop(now + TimeSpan.FromMilliseconds(timeoutMs)))
+            {
+                _options.LogDebug("Profiling is being cut-of after {0} ms because the transaction takes longer than that.", timeoutMs);
+            }
+        }, CancellationToken.None);
+    }
+
+    private bool Stop(DateTimeOffset now)
+    {
+        if (_endTime is null)
+        {
+            lock (_session)
+            {
+                if (_endTime is null)
+                {
+                    _endTime = now;
+                    try
+                    {
+                        _data = _session.Finish();
+                    }
+                    catch (Exception e)
+                    {
+                        _options.LogWarning("Exception while stopping a profiler session.", e);
+                    }
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /// <inheritdoc />
+    public void OnTransactionFinish(DateTimeOffset now)
+    {
+        if (Stop(now))
+        {
+            _options.LogDebug("Stopping profiling collection on transaction finish.");
+        }
+        OnFinish?.Invoke();
+    }
+
+    /// <inheritdoc />
+    public async Task<ProfileInfo?> Collect(Transaction transaction)
+    {
+        if (_data is null || _endTime is null)
+        {
+            _options.LogDebug("Profiling collection cannot proceed because it doesn't seem to have finished properly.");
+            return null;
+        }
+        _options.LogDebug("Starting profile processing.");
+
+        try
+        {
+            _transaction = transaction;
+            using var traceLog = await CreateTraceLogAsync().ConfigureAwait(false);
+
+            if (traceLog is null)
+            {
+                return null;
+            }
+
+            try
+            {
+                if (_cancellationToken.IsCancellationRequested)
+                {
+                    return null;
+                }
+
+                _options.LogDebug("Converting profile to Sentry format.");
+
+                var processor = new TraceLogProcessor(_options, traceLog)
+                {
+                    MaxTimestampMs = (ulong)(_endTime.Value - _startTime).TotalMilliseconds
```

Comment:
This seems like we should be using a stopwatch directly.

Reply:
I think `SamplingTransactionProfiler` should keep an internal stopwatch.  It shouldn't ask for the start/stop times (or another stopwatch) from the transaction.  `DateTimeOffset now` can be removed from the `Start` and `Finish` methods on the interface.  There's really no need (as far as I can tell) for the profiler to care about the date and time-of-day, only the elapsed duration.

I see where the profile's `StartTimeStamp` is serialized, but that can come from the transaction directly.

From https://develop.sentry.dev/sdk/profiles/ 

> Ideally, this will be the same as the field start_timestamp on the transaction so all relative timestamps for transaction and profile can be both from the same reference point and we don't need to correct relative timestamps.